### ISSUE TITLE
Data.List should re-export GHC.List.List

### ISSUE THREAD

**Issue by Kleidukos** [haskell/core-libraries-committee#182]
## Proposal

Data.List should re-export [GHC.List.List](https://hackage.haskell.org/package/base-4.18.0.0/docs/GHC-List.html#t:List)

It is terribly awkward to have 

```haskell
import Data.List qualified as List
import GHC.List (List)
```

## Impact Assessment

I'll fill out this part when I am able to compile the ghc-9.4.4 branch of GHC, for the moment I'm facing hardships with "missing dependencies".

**Comment by treeowl** [comment]
I don't understand what makes this awkward. Also, when do you need `List`?

**Comment by Kleidukos** [comment]
Well, for once, awkwardness levels are influenced by how arm-twisting it is. Today we have the following places for List-related things:

* Data.List
* GHC.List
* GHC.OldList

The first one is the official entry point for List operations in the base library.
The last two are internal modules for GHC, which should not be accessed directly.

Moreover it is highly uncommon to have such a primordial type defined in an internal module, have its operations re-exported in a public-facing module, but not the type itself.

Hence this proposal to solve a very odd mismatch between GHC.List and Data.List

**Comment by treeowl** [comment]
You haven't yet explained the purpose of importing `List` at all, which everything else must lean on.

**Comment by Kleidukos** [comment]
When one imports a type it is usually to use it in a type signature.

**Comment by mixphix** [comment]
FWIW, I'm in favour of exporting this name from `Data.List`. I don't foresee many clashes, other than perhaps type synonyms that would then be obsoleted. Some further justification:

- Many modules are designed with overlapping names, to aid the programmer's memory.
  ```hs
  import Data.List qualified as List (uncons, (\\), ...)
  import Data.Map qualified as Map (fromList, (\\), ...)
  import Data.Set qualified as Set (fromList, (\\), ...)
  import Data.Text qualified as Text (uncons, ...)
  import Data.Vector qualified as Vector (fromList, uncons, ...)
  ```
- Many of these modules are named after their type, which is usually imported unqualified:
  ```hs
  -- import Data.List (List)
  import Data.Map (Map)
  import Data.Set (Set)
  import Data.Text (Text)
  import Data.Vector (Vector)
  ```
- The lack of a *bona fide* prefix name for the type former `[]` is a historical mistake that has only increased the complexity of parsers, and creates a syntactic anomaly at a level where a new programmer is still being introduced to the language. It's a linked list -- it's incredibly simple and useful, but it's not *magic*!
- [Here](https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0475-tuple-syntax.rst#motivation) are some examples of how the current syntax -- "punning" between value- and type-level -- can lead to confusion for newcomers.

**Comment by phadej** [comment]
Isn't this a moral duplicate of https://github.com/haskell/core-libraries-committee/issues/22. i.e. the plan is accepted, but someone needs to write the patches to aid monomorphisation migration of `Data.List`.

**Comment by mixphix** [comment]
I truly don’t understand how this could possibly be interpreted as having anything to do with monomorphization.

**Comment by dixonary** [comment]
As a relative outsider who is also regularly using `List` - including in a new learners context - this is an objectively sensible change and I support it without reservation.

It does not appear to be a duplicate (moral or otherwise) of #22 except in that it also concerns lists.

**Comment by ocharles** [comment]
@phadej perhaps you read this issues as `Data.List` should re-export `GHC.List` (the module), while this issue is really about `Data.List` should re-export `GHC.List.List` (the type).

**Comment by tomjaguarpaw** [comment]
To attempt to fill in some blanks (and I didn't know about this until I just looked), it appears that [`base-4.18.0.0` added `data List a`](https://hackage.haskell.org/package/base-4.18.0.0/docs/GHC-List.html#t:List) which is another way of saying `[a]`. [`base-4.17.1.0` didn't have it](https://hackage.haskell.org/package/base-4.17.1.0/docs/GHC-List.html).

`List` can be preferable to `[]` in some ways, for example

* Avoiding type/term punning
* Avoiding privileged syntax

(I don't know what the relevant status of `[]` and `List` is. I think they are/it is (a) built-in definition(s) but I don't know if one of them counts as a synonym for the other.)

**Comment by jappeace** [comment]
This is a good idea, you've my support. :)

**Comment by chshersh** [comment]
This CLC issue looks relevant:

* https://github.com/haskell/core-libraries-committee/issues/162

I don't like inconsistency. And I'd pretty weirded out by having `List` easily accessible in `base` while `Tuple` doesn't have the same status, even though the existence of both types pursues the same goal (in my understanding).

So my preference is to have the same decision on both of this issues (162 and 182).

**Comment by phadej** [comment]
@ocharles I see. thanks.

In https://github.com/haskell/core-libraries-committee/issues/29 CLC argued that "GHC" stuff could as well be exported just from `GHC.List`. I don't think that GHC proposal 475 will affect ordinary user just yet.

Btw, was there a CLC proposal to add `data List` to `GHC.List` in the first place?

**Comment by int-index** [comment]
I'm not sure how to interpret that last message, but if it's an offer of help with writing CLC proposals, changelog entries, and `@since` annotations, I'll gladly take it. Someone needs to do it.

**Comment by Kleidukos** [comment]
@int-index yeah sure I can add the changelog entry & `@since` annotation as part of the MR that this proposal will have. :)

**Comment by hasufell** [comment]
This might sound nitpicky, but is `List` an internal?

And if so, we should probably have GHC devs clarify, before we start imposing restrictions on them by doing re-exports without asking. 

Although this seems like a no-brainer, we should follow some proper procedure, given the imminent split base proposal.

**Comment by Kleidukos** [comment]
@int-index We spoke briefly of it on Twitter but perhaps could you emit a clarification regarding the status of the List type? Do we risk anything by exposing it?

**Comment by int-index** [comment]
@Kleidukos Yes, it's fine. The new names `List` and `Tuple2`, `Tuple3`, `Tuple4`, etc, are meant to be user-facing. It would be nice to expose the latter from `Data.Tuple`, too.

**Comment by Kleidukos** [comment]
I can retrofit this proposal to include both

**Comment by int-index** [comment]
Doesn't have to be one proposal, that's entirely up to you. I'm just trying to clear up the idea behind [GHC Proposal #475](https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0475-tuple-syntax.rst). Those new names are meant to be user-facing, but they're exported from `GHC.List` and `GHC.Tuple` instead of `Data.List` and `Data.Tuple` to minimize potential breakage (no one has done an impact assessment, as far as I know).

**Comment by mixphix** [comment]
@Kleidukos If you have the capacity to prepare a merge request and impact assessment for exporting both `List` from `Data.List` and `Tuple<n>` from `Tuple`, it would also be much less effort on the CLC's behalf to vote on a single proposal for the matter, while also guaranteeing consistency. :)

**Comment by tek** [comment]
> (no one has done an impact assessment, as far as I know).

I included some of that in https://github.com/haskell/core-libraries-committee/issues/162, for the names that will be exposed for tuples (though not including the modules name `Data.Tuple`. I can amend that and the `List` name.

**Comment by int-index** [comment]
`Tuple<n>` is also in `base`.

**Comment by phadej** [comment]
@int-index, that is a link to `ghc-prim` docs.

**Comment by int-index** [comment]
> that is a link to `ghc-prim` docs.

Good point. So `GHC.List` is in `base` while `GHC.Tuple` is in `ghc-prim`? How strange, I didn't even realize that.

**Comment by tek** [comment]
> namely adding `Unit` and `Tuple<n>` to [`base:Data.Tuple`](https://ghc.gitlab.haskell.org/ghc/doc/libraries/base-4.18.0.0-inplace/Data-Tuple.html)

Those types aren't really exported from that module as far as I can tell, Haddock must have some special treatment for `ghc-prim` that got it confused. Very strange, it also skipped `Tuple16` through `Tuple64`

**Comment by int-index** [comment]
> please add a changelog and `@since` annotations

OK, "please" works way better than "violate an assumption of good faith", so I'll consider doing it.

> the changes you sneaked past CLC 

That concludes my considerations, I have better things to do than shuffle documentation upon your request if you can't even make the request without pairing it with accusations in the same sentence.

I did not sneak anything past anyone, I implemented a section of an _accepted_ GHC Proposal. You know, ordinarily I would expect that someone implementing accepted proposals would get gratitude instead of whatever it is that you are expressing, but I am not easily discouraged, so we don't need to dwell on this. I hope, however, that if someone else contributes to GHC by implementing accepted proposals, they get "thank you" instead of this. (Anyway, they'll certainly get a "thank you" from me).

> Please revert (both in `master` and in `ghc-9.8` branch).

No such action will take place. The changes implement [GHC Proposal #475](https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0475-tuple-syntax.rst). If you want to see those changes reverted, the first step is to amend the GHC proposal, the second step is to implement that amendment (i.e. revert the changes).

> two weeks before I have to escalate

What does that even mean, do you have issues with GHC Proposals getting implemented?

> I'm sorry, but I'm extremely unhappy with this kind of issues surfacing again.

Just a few messages ago I was mostly confused, but by the time I finished typing this response, I certainly share the feeling of unhappiness with you. Perhaps the problem lies in different expectations. Here is how I see the situation:

1. There is an accepted GHC proposal, namely [GHC Proposal #475](https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0475-tuple-syntax.rst).
2. @s-and-witch and I implemented parts of this proposal.
3. We didn't add `@since` annotation and changelog entries; an unfortunate omission, but surely in an open source project anyone could fix this. Perhaps the person who spotted the problem?

**Comment by hasufell** [comment]
> No such action will take place. The changes implement GHC Proposal #475. If you want to see those changes reverted, the first step is to amend the GHC proposal, the second step is to implement that amendment (i.e. revert the changes).

That's unfortunately not how it works. We have talked with @simonpj and @adamgundry during ZuriHac and made some great progress on our mutual understanding of CLC <-> GHC SC.

An accepted GHC proposal that touches base is not guaranteed to be accepted by CLC.

It is the responsibility of the steering committee to contact CLC for such proposals, source their (non-binding) opinions and finally raise an MR and a CLC proposal before merging.

We do want this to synchronize better, which is why that's described in the joint HF tech proposal: https://github.com/Ericson2314/tech-proposals/blob/ghc-base-libraries/proposals/accepted/051-ghc-base-libraries.rst#62ghc-proposals-process

**Comment by int-index** [comment]
> An accepted GHC proposal that touches base is not guaranteed to be accepted by CLC.

That's new. You linked to a fork of `tech-proposals`, so it's only an idea for a new process and it hasn't been accepted by the HF. And I see two major issues with this idea:

1. _Proposal authors_ shouldn't need to interact with two committees instead of one. It is generally fine to have coordination between committees, so I can imagine a system where GHC SC cannot approve a proposal on its own and must contact CLC when a proposal touches `base`. Proposal authors mustn't be bothered with the intricacies of this process, they need to write a single document, submit it on a single platform, and then either have it accepted or rejected, whether by GHC SC alone or by a joint GHC SC + CLC committee.
2. _Proposal implementors_ need a single source of truth for the accepted changes. If the proposal document has been marked accepted, it means that it's ready for implementation. Now, I have no problem whatsoever with involving CLC _before_ a proposal is accepted, but once it's done, it's done. It's quite important to be able to point potential contributors to a document that describes the changes we want to see in GHC and its libraries, and it can't be that parts of this document are actually accepted and parts of it are kind-of-accepted-but-non-binding-and-another-proposal-on-another-platform-is-pending.

So I agree with you that it's important to get CLC involved, but for the sake of proposal authors and proposal implementors, this should be a committee-to-committee interaction, not a person-to-two-committees interaction; and acceptance of a proposal needs to be a single atomic operation instead of having kind-of-accepted documents floating around in the `ghc-proposals` repo.

Not to mention, both 475 and its implementation predate this discussion and this idea for a new process, so accusations of "sneaking" changes past a committee are baseless and quite frankly out of line.

**Comment by hasufell** [comment]
> That's new.

It's not.

Please re-read the charta: https://github.com/haskell/core-libraries-committee#base-package

Nowhere does it say that GHC steering committee gets a free pass. 

> You linked to a fork of tech-proposals, so it's only an idea for a new process and it hasn't been accepted by the HF. 

Not accurate. This proposal has been a joint effort between CLC and GHC steering committee. I'm confused and surprised why you haven't been informed about its progression, since multiple GHC SC members were involved. 

The PR is already up, @Ericson2314 is a co-author and is fixing up wording issues as we continue to discuss it in public and in private: https://github.com/haskellfoundation/tech-proposals/pull/51

And no, we don't expect the HF to vote on it formally. It's not their responsibility  and they have basically already expressed their votum by asking the CLC to get on board: https://github.com/haskell/core-libraries-committee/issues/145

CLC hasn't formally acked it yet, but it's fair to say it seems to enjoy wide consensus so far.

But that doesn't really matter. It's more of a clarification, not a new rule.

Does that answer your questions?

**Comment by s-and-witch** [comment]
> Please revert (both in `master` and in `ghc-9.8` branch). CC @s-and-witch
> 
> @s-and-witch please act on the suggestions within two weeks before I have to escalate.

Please point to the exact lines in `libraries/base` touched by [!9821](https://gitlab.haskell.org/ghc/ghc/-/merge_requests/9821) that you want to be reverted within two weeks before I have to escalate.

**Comment by int-index** [comment]
> I'm confused and surprised why you haven't been informed about its progression, since multiple GHC SC members were involved.

@hasufell I was informed, but apparently I did not pay enough attention. My understanding was that a new package `ghc-experimental` is getting split out of `base`, and I don't particularly care how definitions are distributed across packages (I'd simply add both `base` and `ghc-experimental` to my `build-depends`), so I was not compelled to participate in the discussion. However, I didn't realize that it affected already accepted proposals such as 475 retroactively, nor that it meant that GHC Proposals couldn't be used as a source of truth for accepted changes to GHC and its libraries. Now that you brought this up, I raised this concern on the GHC SC mailing list.

> Does that answer your questions?

Well, I didn't ask any, but your message does clear things up, so thank you. In particular, the fact that "it's more of a clarification, not a new rule" is quite a shift in perspective.

Now, there are two ways to interpret it.
1. It could mean that GHC Proposals can't be taken as a source of truth to guide implementation. This seems to be the interpretation that you are suggesting, but I've already raised my objection with this point of view.
2. It could mean that the GHC Steering Committee should have contacted the Core Libraries Committee before accepting 475, as this proposal changes `base`.

Looking at the mailing list archives, I see that it was I who shepherded 475. I coordinated the discussion between the GHC SC members, but it didn't occur to me to reach out to an external committee. Seems like a similar thing happened with 330, which I also shepherded. I will adjust my practice in light of this discussion, so if I shepherd any proposals that affect `base` in the future, I will get (binding!) CLC feedback before declaring the proposal accepted.

It's not doing anyone any good to have proposals that are accepted-but-not-really.

**Comment by hasufell** [comment]
> It could mean that GHC Proposals can't be taken as a source of truth to guide implementation. This seems to be the interpretation that you are suggesting, but I've already raised my objection to this point of view.

This is correct. Further, @adamgundry explained to us in person at ZuriHac that GHC proposals are quite different from CLC proposals: sometimes a GHC proposal is accepted, but later it is discovered that the implementation is impossible or otherwise problematic and the whole thing is dropped. Regardless of CLC, an accepted GHC proposal is already not guaranteed to make it into GHC.

This is quite different from how CLC operates and might be one of the reasons for tension: CLC requires a working implementation before accepting anything. 

> It could mean that the GHC Steering Committee should have contacted the Core Libraries Committee before accepting 475, as this proposal changes base.

This is what the tech-proposal suggests... consulting CLC beforehand to get a non-binding opinion. 

But it's also fine for GHC SC to accept a proposal and say "we'll have to convince CLC later". It's just a matter of expectations, how you run your process and what your contributors are willing to get into.

----

In that light, I hope you can see why @Bodigrim is unpleasantly surprised and why merging this MR without CLC approval can be interpreted as disrespecting the authority of the CLC committee. 

I think we have to work towards a situation of mutual respect and trust, for the sake of our contributors.

CLC doesn't find any pleasure in blocking progress, but we may have different goals and priorities than GHC SC, for example. And it's important to acknowledge that.

**Comment by s-and-witch** [comment]
> In that light, I hope you can see why @Bodigrim is unpleasantly surprised and why merging this MR without CLC approval can be interpreted as disrespecting the authority of the CLC committee.

There are things that CLC committee can do to avoid such situations in the future, for example add into headers of already accepted GHC proposals cautions like "this proposal partially/fully blocked on CLC, see discussion `[here](https://github.com/haskell/core-libraries-committee/......)`"

I'm not involved in all these administrative processes, I have no time to read all the discussions in all the organizations (GHC proposals, CLC, HF? Something else? IDK) to know what I can implement and what I can't. Do you agree with that no one can/should enforce developers to be involved in all kinds of bureaucratic processes?

If the answer is "yes", then lacking of information about CLC in accepted GHC proposals is a design gap on your side, not on mine. But it turns out that I am a sneaky developer, that (to show disrespect to the authority of the CLC committee, right?) somehow silently merged (after two months of review with 7 other participants of the discussions. Are they partners in my crime?) the patch (that even did not introduce API changes in base) and now I should revert changes (what changes? I did not even touch `base:Data.Tuple` in the patch) to avoid escalation.

**Comment by int-index** [comment]
> Those types aren't really exported from that module as far as I can tell, Haddock must have some special treatment for `ghc-prim` that got it confused. Very strange, it also skipped `Tuple16` through `Tuple64`

@tek Indeed, this seems to be the case. I found this code in `Haddock.Interface.Create`:

```haskell
    -- See Note [Exporting built-in items]
    special_exports
      | mdl == gHC_TYPES  = listAvail <> eqAvail
      | mdl == gHC_PRIM   = funAvail
      | mdl == pRELUDE    = listAvail <> funAvail
      | mdl == dataTupleModule = tupsAvail
      | mdl == dataListModule  = listAvail
      | otherwise         = []
```

And here is the referenced Note:

```
-- Note [Exporting built-in items]
--
-- Some items do not show up in their modules exports simply because Haskell
-- lacks the concrete syntax to represent such an export. We'd still like
-- these to show up in docs, so we manually patch on some extra exports for a
-- small number of modules:
--
--   * "GHC.Prim" should export @(->)@
--   * "GHC.Types" should export @[]([], (:))@ and @(~)@
--   * "Prelude" should export @(->)@ and @[]([], (:))@
--   * "Data.Tuple" should export tuples up to arity 15 (that is the number
--     that Haskell98 guarantees exist and that is also the point at which
--     GHC stops providing instances)
--
```

The fix should be relatively simple. Now that `(~)`, `List`, and `Tuple<n>` are names that _can_ be mentioned in export lists, we can drop the special cases.

**Comment by tek** [comment]
> I found this code in `Haddock.Interface.Create`:

Damn, I grepped for 10 minutes and couldn't find it. Good to know!

**Comment by hasufell** [comment]
> Do you agree with that no one can/should enforce developers to be involved in all kinds of bureaucratic processes?

Yes, we have to document this better for contributors. This is what the tech-proposal tries to do https://github.com/haskellfoundation/tech-proposals/pull/51

So please bear with us.

**Comment by int-index** [comment]
Alright, so I've pushed some code that should address the concerns raised in this thread.

1. [haddock!35](https://gitlab.haskell.org/ghc/haddock/-/merge_requests/35) fixes Haddock rendering of `List`, `Unit`, `Solo`, and `Tuple<n>`. The tuple names are *not* exported from `Data.Tuple`, it's a rendering oddity. I now understand what the request to revert was about, it was based on the incorrect information presented to us by Haddock. Fortunately, there is nothing to revert.
2. [ghc!10835](https://gitlab.haskell.org/ghc/ghc/-/merge_requests/10835) adds the missing `changelog.md` entries and the `@since` annotations that @Bodigrim asked for.  `List` is originally defined in `ghc-prim`, so its `@since` annotation specifies the version of `ghc-prim` rather than `base`. I don't know if there is a way to add a `@since` annotation for a reexport.

I also opened #186 to get approval from the CLC for the `List` export from `GHC.List`.

**Comment by simonpj** [comment]
This confused thread is no one's fault, and no one has acted in bad faith.

Rather, it is a consequence of a structural lack of clarity about how the GHC Steering Committe and the Core Libraries Committee can work together without tripping each other up.

The good news is that we have been working hard on that in the last couple of months, and we have [a detailed plan](https://github.com/Ericson2314/tech-proposals/blob/ghc-base-libraries/proposals/accepted/051-ghc-base-libraries.rst).  The current step is seeking agreement from the GHC SC and CLC.

Returning to this thread, the trouble is that [GHC proposal #475](https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0475-tuple-syntax.rst) predated that plan, and yet is a prime case in point.  Vlad and his colleagues are not to blame for that; they proceeded in good faith.  But we now need to work out what best to do next.

* One possibility is to put the additive bits of  [GHC proposal #475](https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0475-tuple-syntax.rst) into `ghc-experimental`. That does not require CLC approval.
* Another is to make them into additions to `base`.  That requires CLC approval, and means that any future changes (if they turn out to be needed in the light of experience) will also need such approval.

I don't have an opinion about which is best, but I think the best course of action is to make clear which of these you'd like to propose (or a third path), and seek agreement for that path.  I believe you are taking the second path?  But CLC proposals #182 and #185 don't cover the (substantial) additive elements of that MR, do they?  Or are `Tuple2`, `Tuple3` etc only exported from `ghc-prim` and not exposed in `base` at all?  I'm sorry, but I have lost track of what the exact library changes are.  Maybe you can just lay them out for us?

Meanwhile @bodigrim perhaps you can help us by saying what else the CLC needs on this one, beyond the MR that @int-index posted today

@int-index, I know this is frustrating. It's not your fault that you have become a vicitim of shortcomings our process.  But we are where we are, and Julian/Andrew are right: changes to `base` need CLC approval.

We (the GHC SC) will try to do better in future.  My apologies for having failed to do so this time round.

**Comment by tek** [comment]
@simonpj 

>  Or are `Tuple2`, `Tuple3` etc only exported from `ghc-prim` and not exposed in `base` at all? I'm sorry, but I have lost track of what the exact library changes are.

They were intended to be exposed in `GHC.Prelude`, a new module in `base`, for which we asked for approval in #162, which spurred more discussion about split base. 
The most likely scenario, in my view, is for them to be exposed from `ghc-experimental.

However, so far the new `TupleN` definitions are indeed only exported from `ghc-prim`.

**Comment by simonpj** [comment]
> They were intended to be exposed in GHC.Prelude, a new module in base, for which we asked for approval in https://github.com/haskell/core-libraries-committee/issues/162, which spurred more discussion about split base.  The most likely scenario, in my view, is for them to be exposed from `ghc-experimental.

Ah, thank you.  Perhaps the group of people working on this (you, Vlad, etc) can decide what you would prefer, and then re-launch #162, or alterantively withdraw it in favour of `ghc-experimental`?

We defininitely don't want to encourage users to depend on `ghc-prim`!

**Comment by tek** [comment]
I started a discussion on https://gitlab.haskell.org/ghc/ghc/-/merge_requests/8820 .

**Comment by phadej** [comment]
> We defininitely don't want to encourage users to depend on ghc-prim!

Why not? If I want `magicDict` (and I want it often enough), why can't look for a GHC specific thing in `ghc-prim` where it's defined? It's not like it being in `ghc-experimental` will make it any more stable, will it?

In other words, what is the value of re-exporing `TupleN` from somewhere else? I guess we'll see, but IMHO `ghc-prim` is in fact better structured than `GHC.Exts` in `base`, which is just a kitchen sink with all the sharp chef knives mixed up with silver spoons.

**Comment by simonpj** [comment]
> Why not? If I want magicDict (and I want it often enough), why can't look for a GHC specific thing in ghc-prim where it's defined? It's not like it being in ghc-experimental will make it any more stable, will it?

Yes, it will. 

Stuff in `ghc-experimental` is at least a candidate for long term stability.  We wil strive to keep that API stable, albeit not as vigorously as `base`.

But `ghc-prim` (in this new view of things) an entirely internal matter.  It is full of sharp knives.  There is no signal about what functions are stable functios that a user might reasonably rely on, and which are internal functions that might vanish at a moment's notice.

Everyone is welcome to depend on `ghc-prim`, and there are legitmate reasons to do so.   But it is inadvisable to do so without good reason.  If there is a useful function there, which you'd like to be part of GHC's stable API, then please encourage us to put it in `ghc-experimental` or `base`.

You are right about `GHC.Exts`, which predated all this.  Over time, I hope we can give `ghc-experimental` a better structured API.

**Comment by phadej** [comment]
@simonpj naming a package with strong stability guarantees `...-experimental` is not very intuitive.

**Comment by simonpj** [comment]
> @simonpj naming a package with strong stability guarantees ...-experimental is not very intuitive.

That is a fair comment.  Do you have a better suggestion?   It's expermental in the sense that newly-minted GHC Proposal that define new types or functions are likely to expose them here.

Its stability guarantees are not strong; they are stronger than `ghc-prim` (which has none) but weaker than `base`.

**Comment by ocharles** [comment]
I think the name is fine, it just needs to be accompanied by clear prose outline what that stability guarantee is. Presumably it's something like "follows PVP, but we won't necessarily go through deprecation cycles" or something? It seems that people know of some sort of stability spectrum with ghc-prim at one extreme and base at the other. It would be nice to know what one can expect at these various points along this spectrum (that may have already been covered, in which case we can just summarise it and pop it in the documentation/README)

**Comment by simonpj** [comment]
> I think the name is fine, it just needs to be accompanied by clear prose outline what that stability guarantee is

Great. It's all been evolving in the last few months, but once the dust settles I think it would be *really* good idea to document it very explicitly.  Ping ghc-devs if you think that hasn't happened in the next month or three.  (Summer break coming.)

**Comment by s-and-witch** [comment]
IMO, the idea of `ghc-experimental` can be easier replaced with a new pragma like `deprecated`, but for experimental functions/modules. In that case, we can just let experimental functions live in `base` and require CLC proposal only for making them stable.

Similar to this proposal about extensions: https://github.com/ghc-proposals/ghc-proposals/pull/601

By doing things in this way, we will gain advantages that can't be reached (at least all of them) with the new `ghc-experimental` package: 
1) Stable functions can be internally implemented via experimental ones (unless it may change stable function's behavior)
2) Unstable functions can use stable functions defined in `base`
3) No need to know about another new magic standard package that holds "unstable" functions.

**Comment by phadej** [comment]
@ocharles 

> stability spectrum with ghc-prim at one extreme

But `ghc-prim` is extremely stable, I'd say a lot more stable than `base`!

**Comment by phadej** [comment]
> You can't rely on it being stable.

It follows PVP, and I doubt you will rename all the primops just for the sake of renaming.

---

What is the purpose of `ghc-experimental`, is it the staging library for inclusion into `base`?

Will primops be exported from `ghc-experimental`, will primops be included in `base`?

I think these are in conflict.

If `ghc-experimental` is a staging library for stuff which well (hopefully) eventually land in `base`, then primops doesn't belong there, as these will come and go.

And then I think, importing them from `ghc-prim` is **the right thing to do**. If I reach out for a primitive, I know that it may change in the next GHC version.

If on the other hand, `ghc-experimental` is a staging library, but also a library (re)exporting some GHC specific low-level stuff which is not intended to ever be in `base`, then there are two jobs for one library, and that is not great either. That's exactly the problem with current `base`, which tries to serve multiple jobs with conflicting interests.

**Comment by tomjaguarpaw** [comment]
A couple of observations:

1. Discussions about `ghc-experimental` are not particularly relevant to this CLC proposal.
2. Packages created by Team GHC are not the CLC's business in any case. Team GHC can make packages called `ghc-prim`, `ghc-experimental` and `ghc-magic-unicorn` and do whatever they want with them, and none of that impinges (directly) on the CLC.

So I think this discussion has gone way off topic.

**Comment by Kleidukos** [comment]
The following packages would be disrupted by such a change, thanks to serokell's hackage-search and clc-stackage:


acme-grawlix-0.1.0.2
curry-frontend-2.0.0
lorentz-0.15.2
streamly-0.9.0
streamly-core-0.1.0
Mapping-2.0.0.0
Top-1.7
cherry-core-alpha-0.5.0.0
hydrogen-prelude-0.20
kansas-lava-0.2.4.5
liquidhaskell-0.9.0.2.1
local-search-0.0.7
nri-prelude-0.6.0.6
spiros-0.4.2
rebase-1.19

rebase is the only one that raised an error related to `Data.List.List` during the running of clc-stackage.

I will happily submit PRs to these packages.

**Comment by parsonsmatt** [comment]
I'm in favor.

Most of the packages above define `type List = []` anyway.

**Comment by mixphix** [comment]
yes yes yes!

**Comment by hasufell** [comment]
I'm mostly indifferent. The only counter-argument I could see is that beginners might now be exposed to two ways of expressing lists, since it'll be much more prominent in `Data.List`. I don't know what people with more teaching experience have to say about this... @chshersh ?

**Comment by tomjaguarpaw** [comment]
I would be in favour.

**Comment by hellwolf** [comment]
> As a Haskell educator, I did notice that the `[a]` type notation for lists can be confusing to some people, and I'd rather prefer having `List a` instead

Just wanted to mention this, maybe it's still related: https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0475-tuple-syntax.rst

**Comment by Bodigrim** [comment]
From educational perspective, it is a common pedagogical technique to say "let's write our own list `data List a = Nil | Cons a (List a)`". E. g., I think "Programming in Haskell" by @grahamhutton does exactly this, and LYAH is the same, and many others:
* https://discourse.haskell.org/t/how-to-use-this-type-data-list-a-nil-cons-a-list-a/4954
* http://www.cse.unsw.edu.au/~cs1011/04s1/lecture/datatypes-3.pdf
* http://learnyouahaskell.com/making-our-own-types-and-typeclasses#recursive-data-structures
* https://mmhaskell.com/liftoff/data-types
* https://www.haskellforall.com/2021/02/

I'm uneasy that all these resources will be invalidated by exporting `Data.List.List`.

**Comment by dixonary** [comment]
Why would they be invalidated by that? 

* If `List` is not exported from `Prelude` and they don't import `Data.List` then they will still behave as is. 
* If they do import it, then nonprinted resources can adjust to `MyList` or `import Data.List hiding (List)`.
* If there are any *printed* resources who do `import Data.List` unqualified then they'll have been written for an older version of GHC/base which doesn't have this behaviour.

If there are any *printed* resources which `import Data.List` unqualified *and* have no specified GHC that they are written for, then hopefully our modern compiler messages will make the problem obvious to a relative newcomer. If they don't then that's on us.

**Comment by MangoIV** [comment]
I do also agree from what I have seen in students at my uni that promoting less punning is worth more than the possibility of redefining something that already exists; I think the error messages for redefining are pretty readable and easy to solve, even for very beginners; I would even go a bit further and claim that they're one of the first error messages learners see. :D

**Comment by Bodigrim** [comment]
@dixonary the experience is that even non-printed resources are extremely unlikely to get updated.

I mean, this is not a blocker indeed. But I also see only marginal benefit from re-exporting `List` from `Data.List`. As long as we are not looking to deprecate the punning syntax any time soon, I personally don't see much point to promote non-punning one: having two names for the same thing is never good. It's already confusing that because of a Haddock bug, `[a]` and `List a` appear interchangeably in haddocks.

**Comment by Bodigrim** [comment]
@Kleidukos just a reminder that this is still blocked on impact assessment, which requires preparing specific patches, not just listing packages which are potentially affected.

**Comment by Kleidukos** [comment]
@Bodigrim yes! My computer's power supply has problems so I'm waiting for a new one, so that running clc-stackage can be done without problems. :)

**Comment by Bodigrim** [comment]
@Kleidukos if there is no progress within two weeks, I'll close as abandoned.

**Comment by Kleidukos** [comment]
@Bodigrim I should be able to give a proper patch list on Monday! :)

**Comment by Kleidukos** [comment]
Alright, I was actually wrong in my first observation, there are five patches needed:

* [ptr](https://github.com/Kleidukos/ptr/commit/6c9432e1b82893ff28fcdc89fc13769f4a151db3)

* [hpack](https://github.com/Kleidukos/hpack/commit/0b9563c9eef0bedb3a9014f1849fd6413a0133c4)

* [strict-list](https://github.com/Kleidukos/strict-list/commit/ef770990bd52aca8e6893075815867effc961d05)

* [rattletrap](https://github.com/Kleidukos/rattletrap/commit/dbab2d81b67bc02db1a5d002d9dc5e27befeace5)

* [rebase](https://github.com/Kleidukos/rebase/commit/43c9eeb9a1381e906f3cb346774031a4efa6dd9b)

In the last patch I toyed a bit with CPP, which is what should be expected of the final patches that I shall send to these packages once the version of `base` that will hold the new export will be known.

**Comment by Bodigrim** [comment]
Thanks @Kleidukos! Could you also point us to an MR with the proposed change? Sorry if you mentioned it already, I’m from mobile.

**Comment by Kleidukos** [comment]
@Bodigrim Sorry I believe I don't fully understand what you mean, if you are referring to concrete patches for the build problems, they are listed in my above comments (which is what I read I should do before approval, and then submit the patches to the packages _after_ approval). Or perhaps are you thinking of a MR to `head.hackage`? Or did I really misunderstand the [assessment section](https://github.com/haskell/core-libraries-committee/blob/main/PROPOSALS.md#impact-assessments) of the guide? (in which case I am sorry)

**Comment by Bodigrim** [comment]
All good with impact assessment, many thanks :) I'm talking about a draft GHC MR, adding the reexport (just two lines of changes and changelog, nothing fancy).

**Comment by Kleidukos** [comment]
Heavens, apologies for the evening brain. Sure, it's underway. :)

**Comment by Kleidukos** [comment]
MR created: https://gitlab.haskell.org/ghc/ghc/-/merge_requests/11426

**Comment by Bodigrim** [comment]
Dear CLC members, following the non-binding discussion [above](https://github.com/haskell/core-libraries-committee/issues/182#issuecomment-1646224368), let's vote on the proposal to re-export `GHC.List.List` from `Data.List`. The (trivial) MR is available at https://gitlab.haskell.org/ghc/ghc/-/merge_requests/11426 (disregard CI failures, they are inconsequential). The impact assessment is available [above](https://github.com/haskell/core-libraries-committee/issues/182#issuecomment-1723385840): just a few name clashes, which the proposer has patches ready to be submitted (and there are 6+ months until GHC 9.10 to incorporate the change).

@mixphix @hasufell @tomjaguarpaw @parsonsmatt @angerman @velveteer

**Comment by hasufell** [comment]
+1 under the condition that the documentation remarks raised [here](https://github.com/haskell/core-libraries-committee/issues/182#issuecomment-1647918334) are taken care of

**Comment by tomjaguarpaw** [comment]
+1

**Comment by Kleidukos** [comment]
@hasufell Please see https://gitlab.haskell.org/ghc/ghc/-/merge_requests/11426, feedback would be welcome on the topic of documentation.

**Comment by parsonsmatt** [comment]
+1

**Comment by Bodigrim** [comment]
@mixphix @angerman @velveteer just a gentle reminder to vote.

**Comment by velveteer** [comment]
+1

**Comment by mixphix** [comment]
+1

**Comment by Bodigrim** [comment]
+0. I'm generally opposed to have two names for the same thing and reluctant to make sacrifices just for the sake of non-punning Dependent Haskell syntax. But given that it's `List` internally already, and both names resurface in documentation already, embacing the change might be a reasonable option. 

Thanks all, there are enough votes of unconditional support to approve. Further documentation improvements, if any, do not require a CLC vote.

**Comment by Kleidukos** [comment]
Thanks everyone for your patience. :bow:
